blog/2022-11-03-short-circuiting.md

```diff
@@ -0,0 +1,156 @@
+entirely in-place.  The updated values of the row each depend only on the
+previous value, and so they can be computed and written entirely in parallel
+with no risk of data-dependency errors.  Unfortunately, Futhark was not able to
+do so.  The conservative nature of high-level data-race guarantees, such as the
+ones Futhark provide, result in unnecessary memory being allocated and
+extraneous ceopies.
+
+This is further complicated by that fact that, in order to easily express
+complex computations like NW (TODO LINK?), Futhark supports flat slices, which
+make it hard for the compiler to verify that there are no data-dependencies.
+For instance, in our NW benchmark, the main loop boils down to a call like the
+following:
+
+```
+A[W] = map f A[P1] A[P2]
+```
+
+where `W`, `P1` and `P2` are complex slicing operations corresponding to
+staggered blocks.  Even though `f` has no free variables, we still need to
+guarantee that the elements written in one thread (part of `W`) does not overlap
+with the reads (`P1` and `P2`) from any other thread.
+
+# Response
+
+The solution we have implemented in Futhark is called short-circuiting. It works
+on one of Futhark's intermediate representations, `GPUMem`, which is a version
+of the IR with kernel operations and memory optimizations. `GPUMem` looks
+something like this[^1]:
+
+```
+let A_mem: mem = alloc(...)
+let A: [n][m]i64@A_mem -> 0 + {(n:m), (m:1)} = segmap (i < n) {
+    let B_mem: mem = alloc(...)
+    let B: [m]i64@B_mem -> 0 {(m:1)} = segmap (j < m) {
+        let x = i * m + j
+        in x
+    }
+    in B
+}
+```
+
+If you squint a bit, this looks like regular Futhark: The `segmap`s correspond
+to `map` with an explicit index, the arrays have size-types and so on.  But we
+also have a new type, `mem`, an extra `alloc` expression which returns a value
+of type `mem` and arrays have some extra annotations, a memory block and an
+index function. The extra annotations on arrays are used to indicate where in
+memory that array resides and the memory layout of the array.
+
+The short-circuit analysis works by identifying so-called "short-circuit points"
+such as the last statement in this piece of code:
+
+```
+let A_mem: mem = alloc(...)
+let A: [n][m]i64@A_mem -> 0 + {(n:m), (m:1)} = ...
+let B_mem: mem = alloc(...)
+let B: [m]i64@B_mem -> 0 + {(m:1)} = ...
+let A[k] = B
+```
+
+and then trying to prove that it is safe to construct `B` directly in the memory
+of `A` instead of relying on an intermediate buffer.  In the example above, we
+call `A` the destination of the short-circuit attempt, `B` the source and `W` is
+the slice that dictates how `B` should be placed inside.  The result of a
+successful short-circuiting is that `B` is updated with the type `[m]i64@A_mem
+-> k*m + {(m:1)}`, which in turn makes the copy on the last line a no-op.
+
+We have a paper being published in the SC22 proceedings about the details of
```

Comment:
Can't we link to the paper?

Reply:
Yes. I guess we should upload our own version to the futhark-lang website anyway?

Reply:
Yes.  I'll upload both of the papers soon.

Reply:
They are uploaded.